x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanStreamInput.java

```diff
@@ -206,4 +210,21 @@ static void throwOnNullOptionalRead(Class<?> type) throws IOException {
     public NameId mapNameId(long l) {
         return nameIdFunction.apply(l);
     }
+
+    @Override
+    public Attribute attributeFromCache(int id) {
+        if (attributesCache[id] == null) {
+            throw new IllegalArgumentException("Attribute ID not found in serialization cache [" + id + "]");
+        }
+        return attributesCache[id];
+    }
+
+    @Override
+    public void cacheAttribute(int id, Attribute attr) {
+        // IDs are generated and serialized in ascending order
```

Comment:
The fact that this is a comment might show a bit of fragility. Why not treat the IDs opaquely on the receiver end and use a Map here as well? It would mask how the cache is implemented on the emitting end, to some degree (up to it being an int).
BTW, what if it's a negative int? It shouldn't, but it's not checked either.

Reply:
It's mostly for memory footprint and efficiency (an array is much faster and smaller than a map, and apparently we use this cache a lot).
Negative ints should never be generated, I'll add an assert

Reply:
Had the same thought about using a map on both sides; it's only needed for serialization, though (because we need to check object identity), and an array is indeed faster - being fast while occupying the transport layer's resources is important, right? (I think it has an own threadpool that we may be hogging if we're slow?)